nh3_testcubes.py

```diff
@@ -60,33 +112,75 @@ def generate_cubes(nCubes=100, nBorder=1, noise_rms=0.1,
-        Voff2[too_close]=max_Voff+np.max(np.column_stack((Width1[too_close], Width2[too_close])), axis=1)/2.
+    Width1 = np.sqrt(Width1NT + 0.08 ** 2)
+    Width2 = np.sqrt(Width2NT + 0.08 ** 2)
 
-    scale = np.array([[0.2, 0.1, 0.5, 0.01]])
-    gradX1 = np.random.randn(nCubes, 4) * scale
-    gradY1 = np.random.randn(nCubes, 4) * scale
-    gradX2 = np.random.randn(nCubes, 4) * scale
-    gradY2 = np.random.randn(nCubes, 4) * scale
+    Temp = np.array([Temp1, Temp2]).swapaxes(0, 1)
+    Width = np.array([Width1, Width2]).swapaxes(0, 1)
+    Voff = np.array([Voff1, Voff2]).swapaxes(0, 1)
+    logN = np.array([logN1, logN2]).swapaxes(0, 1)
+
+    return nComps, Temp, Width, Voff, logN
+
+
+
+def make_cube(nComps, nBorder, xarr, Temp, Width, Voff, logN, gradX, gradY, noise_rms):
+    # the length of Temp, Width, Voff, logN, gradX, and gradY should match the number of components
+    xmat, ymat = np.indices((2 * nBorder + 1, 2 * nBorder + 1))
+    cube = np.zeros((xarr.shape[0], 2 * nBorder + 1, 2 * nBorder + 1))
+
+    results = {}
+    results['Tmax_a'], results['Tmax_b'] = (0,) * 2
+
+    for xx, yy in zip(xmat.flatten(), ymat.flatten()):
+
+        spec = np.zeros(cube.shape[0])
+
+        for j in range(nComps):
+            # define parameters
+            T = Temp[j] * (1 + gradX[j][0] * (xx - 1) + gradY[j][0] * (yy - 1)) + 5
+            if T < 2.74:
+                T = 2.74
+            W = np.abs(Width[j] * (1 + gradX[j][1] * (xx - 1) + gradY[j][1] * (yy - 1)))
+            V = Voff[j] + (gradX[j][2] * (xx - 1) + gradY[j][2] * (yy - 1))
+            N = logN[j] * (1 + gradX[j][3] * (xx - 1) + gradY[j][3] * (yy - 1))
+
+            # generate spectrum
+            spec_j = ammonia.cold_ammonia(xarr, T, ntot=N, width=W, xoff_v=V)
 
-    params1 = [{'ntot':14,
-                'width':1,
-                'xoff_v':0.0}] * nCubes
-    params2 = [{'ntot':14,
-                'width':1,
-                'xoff_v':0.0}] * nCubes
+            if (xx == nBorder) and (yy == nBorder):
+                Tmaxj = np.max(spec_j)
+                results['Tmax_{}'.format(ascii_lowercase[j])] = Tmaxj
+
+            # add each component to the total spectrum
+            spec = spec + spec_j
+
+        cube[:, yy, xx] = spec
+        if (xx == nBorder) and (yy == nBorder):
+            Tmax = np.max(spec)
+            results['Tmax'] = Tmax
+
+    # should we add noise independently on individual spectrum to avoid correlated noise across pixels?
+    cube += np.random.randn(*cube.shape) * noise_rms
+    results['cube'] = cube
+    return results
+
+
+
+def write_fits_cube(cube, nCubes, nComps, i, logN, Voff, Width, Temp, noise_rms,
+                    Tmax, Tmax_a, Tmax_b, lineID='11', output_dir='random_cubes'):
+    """
+    This places nCubes random cubes into the specified output directory
+    """
+    if not os.path.isdir(output_dir):
+        os.mkdir(output_dir)
+
+    logN1, logN2 = logN[0], logN[1]
```

Comment:
This may cause problems if we consider nComp > 2.  Not really important now though.